### PR TITLE
storage: remove mutex

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang @disksing @zhangjinpeng1987 PTAL

### CODE REVIEW

File: src/storage/engine/mod.rs [hunk 1]
```diff
@@ -73,6 +73,9 @@ pub trait Engine: Send + Sync + Debug {
     fn delete_cf(&self, ctx: &Context, cf: CfName, key: Key) -> Result<()> {
         self.write(ctx, vec![Modify::Delete(cf, key)])
     }
+
+    /// Create a share Engine pointer.
+    fn clone(&self) -> Box<Engine + 'static>;
```
Comment (disksing) [inline src/storage/engine/mod.rs]: How about defining Engine as `trait Engine: Send + Debug + Clone` instead?

Comment (BusyJay) [inline src/storage/engine/mod.rs]: No, `Clone` trait will prevent creating a `Box<Engine>`.

Comment (disksing) [inline src/storage/engine/mod.rs]: Can we just return an `Engine` here?

Comment (BusyJay) [inline src/storage/engine/mod.rs]: Engine is not `Sized`, so here either return a trait object or generic. And return Self will face the same problem when using `Clone`.

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**